FLIGHT_DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,442 @@
+object is used across flights, each flight would be updating the number of retries.
+
+The reason that you can share a RetryRule object within a flight is that RetryRules have
+an `initialize()` method that resets the state. The Flight object always initializes the
+RetryRule object at the start of executing the `do()` and `undo()` methods of a Step.
+
+Every step has a RetryRule object associated with it. If the rule is not specified, the
+`RetryRuleNone` is used. RetryRuleNone has no state and does no retries. It always
+returns: "do not retry".
+
+### How to Retry <a name="retryhow"></a>
+There are two ways from with a step to request that the step be retried.
+ 1. Throw an exception that derives from `RetryException`. When flight execution catches
+ that exception, it will attempt a retry.
+ 1. Return a StepResult of `STEP_RESULT_FAILURE_RETRY`.
+
+It is our experience that having the step perform its own exception handling and return
+the explicit `STEP_RESULT_FAILURE_RETRY` is more readable. It also keeps service
+exceptions separate from Stairway exceptions.
+
+Remember that simply requesting a retry and actually retrying are two separate
+things. Requesting retry asks Stairway to check the RetryRule to see if there are any more
+retries left. If the rule says, "no more retries", then the exception returned in the
+StepResult is used as the result exception of the flight.
+
+### Retry After Restart <a name="retryafterrestart"></a>
+Retrying is only in-memory. RetryRule state is not retained across a failure/recovery.
+
+Imagine that Stairway is running a step of a flight that uses a retry rule. The rule will
+try 10 times and then give up. Stairway has retried 9 times and then the server fails.
+When the flight is recovered, the step will be restarted and the retry rule counter will
+start at 0 once again.
+
+## Flight Construction <a name="flightconstruction"></a>
+
+The Stairway caller never constructs a Flight object. When you call Stairway to run a flight,
+you provide the input parameters and the class name. Stairway does the actual
+construction. That is because Stairway must be able to re-construct the flight after a
+failure/recovery from the data stored in the Stairway database.
+
+### Constructors Should... <a name="constructorsshould"></a>
+
+Your flight constructor code should construct and configure the steps and retry rules for
+the flight. The constructor might include:
+ - adding different steps based on input parameters
+ - extracting input parameters and providing them as constructor parameters to the steps
+
+### Constructors Should NOT... <a name="constructorsshouldnot"></a>
+
+The constructor **should not**:
+ - do database lookups
+ - access external services
+
+There are two reasons for those constraints. First, flight construction is done in
+Stairway on the recovery path. If a flight cannot be reconstructed, Stairway will
+fail to initialize. That can leave your service in a failed state.  
```

Comment:
Is this still true/desirable? Seems like we could mark these as `FATAL` or worse and fire an alert for a human to look at while letting the service continue to start up.

Reply:
The constraints are true, regardless of whether they stop Stairway from initializing.